### ISSUE TITLE
Schema builder is mutating schemas

### ISSUE THREAD

**Issue by jsumners** [fastify/fastify#1767]
## 🐛 Bug Report

After a route is added a function `afterRouteAdded` is invoked which in turn invokes `validation.build` => `schemas.getSchemaAnyway` (via `schema.resolveRefs`) to build out schemas. This `getSchemaAnyway` function is injecting new properties into the schema object before passing the schema to the defined schema compiler. If said schema compiler does not recognize these newly added properties then it can fail to return a compiled schema that will validate as expected.

https://github.com/fastify/fastify/blob/9b233e8b803bc6e4767d0cc865df4af1d074273c/lib/schemas.js#L121-L126

## To Reproduce

Let's say you use Joi schemas and that your have a custom Joi schema compiler that recognizes some internal extensions to Joi. If you define a route like:

```js
const route = {
  path: '/foo/:an_id',
  schema: {
    params: {
      an_id: joi.number()
    }
  },
  handler(req, res) {
    assert.deepStrictEqual(req.params, {an_id: 42})
    res.send({hello: 'world'})
  }
}
```

Then a request to `/foo/42` will fail the assertion because the custom compiler will return a schema that does not recognize any properties in the `param` schema due to the extra data being inserted by `getSchemaAnyway`.

## Expected behavior

`getSchemaAnyway` should not mutate schema definitions it does not recognize.

## Your Environment

- *node version*: 10
- *fastify version*: >=2.6.0
- *os*: Mac

**Comment by jsumners** [comment]
Replication:

```js
'use strict';

const joi = require('@hapi/joi');
const fastify = require('fastify')();

function schemaCompiler(schema) {
  return (data, opts) => joi.validate(data, schema);
}

fastify.setSchemaCompiler(schemaCompiler);

fastify.route({
  path: '/foo/:an_id',
  method: 'GET',
  schema: {
    params: {
      an_id: joi.number()
    }
  },
  handler(req, res) {
    res.send(req.params);
  }
});

fastify.inject({ url: '/foo/42' }, (err, result) => {
  console.log(result.payload);
});
```

**Comment by jsumners** [comment]
Note that this reproduction script works under Fastify v1.14.6 without any changes.

**Comment by jsumners** [comment]
Commit 4c4f77517c8ecc1dd7cb63215dedfcd862674f8b is the one that introduces the bug.

**Comment by jsumners** [comment]
Fixed in #1768
